### ISSUE TITLE
New > File Errors for Java Gradle Project

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#805]
When I was testing today with the latest master build on macOS, I noticed some problems with New > File for a Java Gradle project.
1. Create a new Java Gradle project
1. Right-click on the project and choose New > File
1. BUG:  An error triggers on the console
1. BUG:  The project is not selected properly in the New File tree
1. Select the project.  BUG:  an error triggers on the console
1. Enter "TestNew.txt" as the name.  BUG:  interface shows "Class files allowed to create inside source folder only.".  This rule should only be applied for Java class and interface files
1. Choose the src directory.  The same error triggers on the console
1. Try to expand the src directory.  BUG:  there is no subdirectory in this tree.

I did not see these problems in a Flex Desktop project, but it could affect other project types.

Here is the error when opening the prompt:

    : TypeError: Error #1009
    : 	at components.popup.newFile::NewFilePopup/onNewFileCreationComplete()
    : 	at components.popup.newFile::NewFileBase/___NewFileBase_ResizableTitleWindow1_creationComplete()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at mx.core::UIComponent/dispatchEvent()
    : 	at mx.core::UIComponent/set initialized()
    : 	at mx.managers::LayoutManager/validateClient()
    : 	at mx.core::UIComponent/validateNow()
    : 	at mx.managers::PopUpManagerImpl/centerPopUp()
    : 	at mx.managers::PopUpManager$/centerPopUp()
    : 	at actionScripts.plugin.templating::TemplatingPlugin/openNewComponentTypeChoose()
    : 	at actionScripts.plugin.templating::TemplatingPlugin/handleNewTemplateFile()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at components.views.project::TreeView/handleNativeMenuItemClick()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at mx.core::UIComponent/dispatchEvent()
    : 	at actionScripts.ui.renderers::FTETreeItemRenderer/redispatchNew()
    : 	at flash.display::NativeMenu/display()
    : 	at flash.ui::ContextMenu/display()
    : 	at flash.display::InteractiveObject/onContextMenuEvent()
    :
    : Click here to Report a Bug

The initial state of the tree looks like this:
![image](https://user-images.githubusercontent.com/24532731/114791849-ce59d080-9d4c-11eb-90ef-8f477b2611af.png)

Here is the error for selecting a directory in the tree:

    : Error #1009
    : TypeError: Error #1009
    : 	at components.popup.newFile::NewFilePopup/onTreeItemClicked()
    : 	at components.popup.newFile::NewFilePopup/__tree_itemClick()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at mx.core::UIComponent/dispatchEvent()
    : 	at mx.controls.listClasses::ListBase/mouseClickHandler()
    : 	at mx.controls::Tree/mouseClickHandler()

Here is what the New File dialog looks like at the end of the test procedure:
![image](https://user-images.githubusercontent.com/24532731/114791837-ca2db300-9d4c-11eb-8e12-e9ed36dca17c.png)

**Comment by JoelProminic** [comment]
I see the problem is fixed in the base project directory, but not in any subdirectories.
1. Create/Open a Java Gradle project
2. Add a directory called "test" with Right Click > New > Folder
3. Right-click on the project and choose New > File
4. Select the "test" directory (also triggers for "src"
5. Type "Test.txt" for the file name
6. BUG: interface shows "Class files allowed to create inside source folder only.". 


This issue was not resolved:  "8. Try to expand the src directory. BUG: there is no subdirectory in this tree.".  I see that if I expand src/main/java in the project sidebar before opening the New File prompt, then I can explore the expanded directories.  @rat-moonshine, could this be related to the tree loading changes for #756?


Also, the error text should be changed from:

> Class files allowed to create inside source folder only.

To:

> Class files may only be created inside a source folder.

**Comment by rat-moonshine** [comment]
Few inputs from my tests (I've downloaded the last nightly 'master' build):

1. Create/Open a Java Gradle project
2. Opened new folder creation option by right-clicking on its "src" folder
3. Co-incidentally I right-clicked on the opened tree in the new folder creation dialog and it throws following error:

```
: Click here to Report a Bug
: Error #1009
: TypeError: Error #1009
: 	at actionScripts.ui.renderers::FTETreeItemRenderer/updateOverMultiSelectionOption()
: 	at flash.display::NativeMenu/display()
: 	at flash.ui::ContextMenu/display()
: 	at flash.display::InteractiveObject/onContextMenuEvent()
:
: Click here to Report a Bug
```

I remember I stopped right-clicking option in new file dialogs, maybe I forgot to have the change added int he new folder dialog - I'm going to look into that.

From @JoelProminic ,

> 4. Select the "test" directory (also triggers for "src"
> 5. Type "Test.txt" for the file name
> 6. BUG: interface shows "Class files allowed to create inside source folder only.".

The dialog determines the source folder based on the project's [sourceFolder](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/valueObjects/ProjectVO.as#L55) information to a project. Based on that it also adds a "S" icon to the folder in project tree-view, i.e.
![image](https://user-images.githubusercontent.com/24370870/115664021-f48bfb80-a35e-11eb-842c-c8a10ce6b535.png)

I don't see any "S" icon for Gradle project so I can assume the `sourceFolder` not being set and also affects in the class-file creation policy in a new class-file creation dialog. One need to review if the `sourceFolder` can be populate during a Gradle or other Java project imports.

**Comment by rat-moonshine** [comment]
> This issue was not resolved: "8. Try to expand the src directory. BUG: there is no subdirectory in this tree.". I see that if I expand src/main/java in the project sidebar before opening the New File prompt, then I can explore the expanded directories. @rat-moonshine, could this be related to the tree loading changes for #756?

I can try re-looking into this as well.

**Comment by rat-moonshine** [comment]
> I remember I stopped right-clicking option in new file dialogs, maybe I forgot to have the change added int he new folder dialog - I'm going to look into that.

Should have fixed now.

**Comment by rat-moonshine** [comment]
> > This issue was not resolved: "8. Try to expand the src directory. BUG: there is no subdirectory in this tree.". I see that if I expand src/main/java in the project sidebar before opening the New File prompt, then I can explore the expanded directories. @rat-moonshine, could this be related to the tree loading changes for #756?
> 
> I can try re-looking into this as well.

This should be fixed now.

**Comment by piotrzarzycki21** [comment]
> I don't see any "S" icon for Gradle project so I can assume the `sourceFolder` not being set and also affects in the class-file creation policy in a new class-file creation dialog. One need to review if the `sourceFolder` can be populate during a Gradle or other Java project imports.

@rat-moonshine I looked into *..javaproj file and I see:
```

  <classpaths>
    <class path="src/main/java"/>
  </classpaths>
```
I'm assuming that s is showing when this is filled up. Am I right ?

**Comment by rat-moonshine** [comment]
The "S" in tree-view displays on the folder path set in `sourceField` property, unless I'm forgetting something..

**Comment by piotrzarzycki21** [comment]
I looked into object of Java Gradle Project and sourceFolder is setup: 

![source_folder](https://user-images.githubusercontent.com/24554795/115696578-1edfb800-a363-11eb-88da-feb641e2cb67.png)

**Comment by rat-moonshine** [comment]
Okay. Can you please leave this for sometime unless you want to continue on investigation, else I shall take a look a few hour's later on how the `FileWrapper.isSourceFolder` generates - my initial memory on this told me it's rely on `sourceField` property, but I see there could be more or I forgot something.

**Comment by piotrzarzycki21** [comment]
Yep I can leave it to you for now.

**Comment by rat-moonshine** [comment]
For a startup, the `sourceFolder` still the valid property to determine placing the "s" icon by a node in tree-view. Since the changes into #756 the generation of the icon on UI became slightly tricky. I assume the "s" icon used to show properly before that even for the Java projects. 

I think I want to look into this taking a little more. I have raised a separate issue #814 . 

However, @piotrzarzycki21 since the `sourceFolder` is available for the Gradle/Java project, and when @JoelProminic reported that he's getting "Class files allowed to create inside source folder only." - Joel, please make sure you creates the class-files under `src/main/java` - because that is how it is strictly constructed in `JavaImporter` [here](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/plugin/java/javaproject/importer/JavaImporter.as#L113). I see the error disappears once I select `java` under `src/main`.

**Comment by rat-moonshine** [comment]
I think I have fixed "Class files allowed to create inside source folder only." error across all the class-file creation popups.

However, the New File creation popup still has problem. It has many conditional states because the same popup uses for few other functionalities as well, i.e. XML, duplicate-file. I need to spend some time to have this fixed.

**Comment by rat-moonshine** [comment]
Okay. I tried to fix the "Class files allowed to create inside source folder only." in New File dialog. This covers following things:
- Creation of generic files in any folder of a project
- Keep restricting "Duplicate File" feature for known class-files, to project's source-folder only. However, one can duplicate any other generic type of files at any place in the project

**Comment by JoelProminic** [comment]
Testing with a Java Gradle project, I see that I can create .txt files in any directory, and .java files only in src/main/java.   It sounds like this requires more testing for other project types, though.

I noticed another issue with the project tree loading.  It may make more sense to have a separate issue for this:
1. Create or reopen a Java Gradle project.  At this point, the src directory should be collapsed
2. Right-click > New > File
3. Expand and select src/main/java in the prompt
4. Create a text file
5. BUG:  In the project sidebar, src/main/java is open, but the only file is the .txt file you just created
6. Refresh src/main/java.  Now the other files are visible.

**Comment by JoelProminic** [comment]
Also, I'd still like to update the error text while we are thinking about it.  From 

> Class files allowed to create inside source folder only.

To:

> Class files may only be created inside a source folder.

**Comment by rat-moonshine** [comment]
> I noticed another issue with the project tree loading. It may make more sense to have a separate issue for this:
> 
> Create or reopen a Java Gradle project. At this point, the src directory should be collapsed
> Right-click > New > File
> Expand and select src/main/java in the prompt
> Create a text file
> BUG: In the project sidebar, src/main/java is open, but the only file is the .txt file you just created
> Refresh src/main/java. Now the other files are visible.

and 

> Also, I'd still like to update the error text while we are thinking about it. From

Both should be fixed now. Please, check.

**Comment by JoelProminic** [comment]
I confirmed both issues are fixed on macOS and Windows.   I think all of the problems are resolved for this issue, though #814 is still pending.
